app/1.0.x/plugin-development/custom-entities.md

```diff
@@ -129,76 +129,105 @@ To see a real-life example, give a look at the [Key-Auth plugin migrations](http
+  It supports **composite keys**, even if most Kong core entities currently use UUID named
+  `id`. If you are using Cassandra and need a composite key, it should have the same
+  fields as the *partition key*.
+* `endpoint_key`: An optional string, which would be the name of a field used as an
+  alternative identifier on the Admin API. On the example above, `key` is the endpoint_key. This
+  means that a credential with `id = 123` and `key = "foo"` could be referenced as
+  `/keyauth_credentials/123` and `/keyauth_credentials/foo`.
+* `cache_key`: An optional array of field names. It is used for generating the `cache_key`, a
+  string which must unequivocally reference the entity. A unique field, like `key` in your example,
+  is usually good candidate, but sometimes it's useful to combine more than 1 field.
+
+And some notes regarding fields:
+
+* The `typedefs` variable (obtained by requiring `kong.db.schema.typedefs`) is a table containing
+  [a lot of useful type definitions and aliases](https://github.com/Kong/kong/blob/master/kong/db/schema/typedefs.lua),
+  including `typedefs.uuid`, the most usual type for the primary key, and `typedefs.auto_timestamp_s`,
+  for `created_at` fields.
+* Schemas support the following scalar types: `string`, `integer`, `number` and `boolean`. These types
+  will need to be backed by fields of appropriately similar types in the database. Compound types like Arrays, Records,
+  or Sets are also supported.
+* Foreign relationships can be specified using the `foreign` type on a field, as exemplified by the
+  `consumer` field in the previous example. The `reference` attribute *must* be the name of an existing
+  schema. This means that if a schema B has a foreign key pointing to schema A, then A needs to be loaded
+  before B.
+* The `on_delete = "cascade"` attribute is specific to foreign relationships, and dictates what happens
+  with the "children" when a "parent" entity is removed. These need to be enforced via migrations in PostgreSQL,
+  but not in Cassandra.
+* Default values can be set with the `default` attribute. When a field with a default value is not set on
+  insert, the value is set via Lua, never by the underlying database. It is thus not recommended to set
+  any default values on the databases themselves, for example on migrations.
+* `required = true` works as expected. The DAO will throw an error when attempting to insert an entity
+  lacking required fields.
+* `unique = true` will mark fields so that they can not have the same value in two entities (note
+  that this might require special handling via migrations)
+* `auto = true` will:
+  * replace `null` values with a random UUID when inserting fields of type `uuid`
+  * replacing `null` values with a random string when inserting fields of type `string`
+  * adding the current time when appropriate in fields named `created_at` or `updated_at`
+
+To learn more about schemas, see:
```

Comment:
Before pointing to referential code, we should dissect the above example schema given with the keyauth_credential `daos.lua` file, and elaborate on some interesting bits it has (maybe by way of inlined Lua comments?).

Then, we can set expectations as to what the developers will find in those resources (*not* documentation, but code to read, understand, and reuse, which may be deceptive):

```
A great resource of examples of schema definitions is Kong's source code itself:

* Kong's [bundled plugins daos.lua] files contain custom entities schema definitions that follow 
  the approach documented in this guide. 
* [the core entities schemas] will give you insight to complex entity definitions you can study and reuse
  in your plugins' entities.
* the [typedefs.lua] file contains a list of all bundled types definition you may use for your entity's fields.
```

Reply:
Re: the custom DAO, I have expanded the docs with a new section for it.

I have also added text around the provided links so they are not just links.